core/amber/src/main/python/core/runnables/test_batch_operator.py

```diff
@@ -0,0 +1,861 @@
+                    ),
+                ),
+            )
+        )
+
+        reraise()
+
+    @pytest.mark.timeout(5)
+    def test_dp_thread_can_process_batch_median(
+            self,
+            mock_udf,
+            mock_link,
+            mock_receiver_actor,
+            mock_controller,
+            input_queue,
+            output_queue,
+            mock_data_elements,
+            dp_thread,
+            mock_update_input_linking,
+            mock_add_partitioning,
+            mock_end_of_upstream,
+            mock_query_statistics,
+            mock_batch,
+            command_sequence,
+            reraise,
+    ):
+        dp_thread.start()
+
+        # can process UpdateInputLinking
+        input_queue.put(mock_update_input_linking)
+
+        assert output_queue.get() == ControlElement(
+            tag=mock_controller,
+            payload=ControlPayloadV2(
+                return_invocation=ReturnInvocationV2(
+                    original_command_id=command_sequence,
+                    control_return=ControlReturnV2(),
+                )
+            ),
+        )
+
+        # can process AddPartitioning
+        input_queue.put(mock_add_partitioning)
+        assert output_queue.get() == ControlElement(
+            tag=mock_controller,
+            payload=ControlPayloadV2(
+                return_invocation=ReturnInvocationV2(
+                    original_command_id=command_sequence,
+                    control_return=ControlReturnV2(),
+                )
+            ),
+        )
+
+        # can process a InputDataFrame
+        for i in mock_data_elements:
+            input_queue.put(i)
+
+        output_data_element = []
+        for i in range(50):
+            output_data_element.append(output_queue.get())
+
+        input_queue.put(mock_query_statistics)
+        assert output_data_element[0].tag == mock_receiver_actor
+        assert isinstance(output_data_element[0].payload, OutputDataFrame)
+        data_frame: OutputDataFrame = output_data_element[0].payload
+        assert len(data_frame.frame) == 1
+        assert data_frame.frame[0] == Tuple(mock_batch[0])
+
+        # can process QueryStatistics
+        assert output_queue.get() == ControlElement(
+            tag=mock_controller,
+            payload=ControlPayloadV2(
+                return_invocation=ReturnInvocationV2(
+                    original_command_id=1,
+                    control_return=ControlReturnV2(
+                        worker_statistics=WorkerStatistics(
+                            worker_state=WorkerState.RUNNING,
+                            input_tuple_count=57,
+                            output_tuple_count=50,
+                        )
+                    ),
+                )
+            ),
+        )
+
+        input_queue.put(mock_end_of_upstream)
+        for i in range(7):
+            output_data_element.append(output_queue.get())
```

Comment:
swap the order. you need to sync the output before you can test any statistics. so do line 403 first before you do line 377 and line 385

Reply:
fix